Faculty carousel

singingwolfboy opened the pull request:
Fixes #923. Pulls faculty information from database and renders it in Django Templates. Then that information is parsed and passed to [react-slick](https://github.com/akiran/react-slick) for a nice image carousel effect.

noisecapella commented on cms/templates/cms/program_page.html:
```diff
@@ -104,11 +104,37 @@
     </div>
   </div>
 
+  {# if page.faculty_members.count > 0 #}
```
Just curious, what's the difference between `{%` and `{#` here?

singingwolfboy replied:
`{#` is a comment in Django templates -- this logic was commented-out for testing. Now that the PR is ready for review, it is no longer commented out.

noisecapella commented on static/js/faculty_carousel.js:
```diff
@@ -0,0 +1,64 @@
+    } else {
+      nameStr = this.props.name
+    }
+    let style = {'backgroundImage': 'url('+this.props.imageUrl+')'};
```
Do we need to escape `imageUrl` here?

noisecapella replied:
Nevermind, not ready for review :)

singingwolfboy replied:
I _believe_ that React will take care of the escaping here, but I'm not sure. Can you give me an example of a string that will cause an error, if not properly handled/escaped?

noisecapella replied:
If the url had a `)` in it, you might get `url())`. I think we can just assume the URL is encoded properly before we get it in this component.

singingwolfboy commented on webpack.config.shared.js:
```diff
@@ -25,8 +26,7 @@ module.exports = {
       },  // to transform JSX into JS
       {
-        test: /\.(svg|ttf|woff|woff2|eot)$/,
-        exclude: /node_modules/,
```
I needed to remove this `exclude` line to get the Slick theme to load properly with webpack. I talked about this with @aliceriot, and we both think that _all_ of these `exclude:` lines are unnecessary. Should we maybe remove them in a future pull request?

noisecapella replied:
Sounds good with me

noisecapella commented on static/js/faculty_carousel.js:
```diff
@@ -0,0 +1,89 @@
+import React from 'react';
```
Can you add `// @flow` as the first line of this file to enable flow on it?

singingwolfboy replied:
Done

noisecapella commented on cms/templates/cms/program_page.html:
```diff
@@ -104,11 +104,37 @@
+      {% if page.faculty_description|length > 0 %}
+        <p class="description">{{ page.faculty_description }}</p>
+      {% endif %}
+      <div id="faculty-carousel" class="single-column-wide">
```
A future version of `ReactDOM.render` may leave the HTML below in place so the behavior of replacing the contents of the container is not guaranteed. See the red box: https://facebook.github.io/react/docs/top-level-api.html#reactdom.render

Instead can we populate a JS array with the faculty data? This would also make `FacultyCarousel` easier to test

noisecapella commented on static/js/faculty_carousel.js:
```diff
@@ -0,0 +1,89 @@
+  }
+  render() {
+    let nameStr;
+    if (this.props.title) {
```
Can you use destructuring here: `const { name, title, bio } = this.props;` for example

singingwolfboy replied:
Done

noisecapella commented on static/js/faculty_carousel.js:
```diff
@@ -0,0 +1,89 @@
+
+class FacultyCarousel extends React.Component {
+  props: {
+    faculty:  array,
```
Please change `array` to `Array<Object>`

singingwolfboy replied:
Done

noisecapella commented on static/js/faculty_carousel.js:
```diff
@@ -0,0 +1,89 @@
+}
+
+
+class FacultyCarousel extends React.Component {
```
Can we move this class into a separate file in `static/js/components`?

singingwolfboy replied:
Done

noisecapella commented on static/js/faculty_carousel.js:
```diff
@@ -0,0 +1,89 @@
+
+
+// define a generic `map` function, that can be used on a NodeList
+let map = (...fnargs) => {
```
Why not `Array.from(nodelist).map`?

singingwolfboy replied:
Switched to `Array.from(nodelist).map`

noisecapella commented:
Is there a way to get the dots beneath the carousel to show up?

noisecapella commented:
Can you change the background color of the carousel to match the mockups? My color picker says it's `#0c0c0c`

noisecapella commented:
With two items in the carousel I'm seeing some weird behavior. If I swipe left twice then there is nothing being displayed anymore. If I then swipe left again it goes back to the first item. Can you investigate this?

singingwolfboy commented:
Alright, I've switched over from passing the data via HTML to passing the data via JSON. However, there are a few unsolved problems here:

* The JSON only provides the URL to the full-size image, so if the image is HUGE, the browser will have to download the whole thing and then scale it down. I would like to make the image serializer display multiple URLs for multiple sizes, but I can't figure out the API that Wagtail uses to generate those URLs.
* I switched from a CSS `background-image` attribute to an HTML `<img>` tag, which is better for accessibility since the image contains useful information rather than just being meaningless decoration. However, this `<img>` tag needs to have an `alt` attribute to be accessible. We'll need to add a field in the database for this, and allow CMS users to define it.

Nonetheless, these seem like things that can be cleaned up in the future.

noisecapella commented:
Sounds good, let me know when you want me to take another look

noisecapella commented on cms/models.py:
```diff
@@ -26,6 +26,12 @@ def programs_for_sign_up(programs):
 
+def faculty_for_carousel(faculty):
+    """formats faculty info for the carousel"""
+    from cms.serializers import FacultySerializer
```
Can you move this import to the top of the file?

singingwolfboy replied:
Unfortunately, no. That would cause a circular import problem. (`serializers.py` importing `models.py` importing `serializers.py`...)

noisecapella replied:
Sounds good

noisecapella commented on cms/serializers.py:
```diff
@@ -0,0 +1,28 @@
+
+
+class ImageSerializer(serializers.ModelSerializer):
+    "Serializer for Wagtail Image objects."
```
Can you change the docstrings to triple `"` here and below?

singingwolfboy replied:
Done

noisecapella commented on cms/serializers_test.py:
```diff
@@ -0,0 +1,21 @@
+        faculty = FacultyFactory.create()
+        result = FacultySerializer().to_representation(faculty)
+        assert result['name'] == faculty.name
+        assert result['short_bio'] == faculty.short_bio
```
Can you assert `image` and `title` too?

singingwolfboy replied:
Done

noisecapella commented on static/js/components/FacultyCarousel.js:
```diff
@@ -0,0 +1,69 @@
+import Slider from 'react-slick';
+
+
+class FacultyTile extends React.Component {
```
Can you move this to its own file too?

singingwolfboy replied:
Done

noisecapella commented:
Can you update the help text of `ProgramFaculty.image` to specify a minimum pixel resolution for people to upload? Wagtail will not upscale images, only downscale or crop, so they can't upload an image that's too small.

After that feel free to merge if tests pass :+1:
